roxctl/image/sbom/sbom.go

```diff
@@ -0,0 +1,123 @@
+package sbom
+
+import (
+	"bytes"
+	"encoding/json"
+	"io"
+	"net/http"
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+	imageUtils "github.com/stackrox/rox/pkg/images/utils"
+	"github.com/stackrox/rox/pkg/utils"
+	"github.com/stackrox/rox/roxctl/common"
+	"github.com/stackrox/rox/roxctl/common/environment"
+	"github.com/stackrox/rox/roxctl/common/flags"
+	"github.com/stackrox/rox/roxctl/common/util"
+)
+
+const (
+	imageSBOMAPIPath = "/api/v1/images/sbom"
+)
+
+// Command sends a request to Central to generate an SBOM from an image scan.
+func Command(cliEnvironment environment.Environment) *cobra.Command {
+	imageSBOMCmd := &imageSBOMCommand{env: cliEnvironment}
+
+	c := &cobra.Command{
+		Use:   "sbom",
+		Short: "Generate an SBOM from an image scan.",
+		Long:  "Generate an SBOM from an image scan. Optionally, force a rescan of the image. You must have write permissions for the `Image` resource.",
+		RunE: util.RunENoArgs(func(c *cobra.Command) error {
+			if err := imageSBOMCmd.construct(c); err != nil {
+				return err
+			}
+
+			return imageSBOMCmd.GenerateSBOM()
+		}),
+	}
+
+	c.Flags().StringVarP(&imageSBOMCmd.image, "image", "i", "", "Image name and reference. (e.g. nginx:latest or nginx@sha256:...)")
+	c.Flags().BoolVarP(&imageSBOMCmd.force, "force", "f", false, "The --force flag ignores Central's cache for the scan and forces a fresh re-pull from Scanner")
+	c.Flags().StringVar(&imageSBOMCmd.cluster, "cluster", "", "Cluster name or ID to delegate image scan to")
+
+	utils.Must(c.MarkFlagRequired("image"))
+	return c
+}
+
+// imageSBOMCommand holds all configurations and metadata to generate an SBOM.
+type imageSBOMCommand struct {
+	image   string
+	force   bool
+	cluster string
+
+	env    environment.Environment
+	client common.RoxctlHTTPClient
+
+	// The HTTP request body to send to Central.
+	requestBody []byte
+}
+
+// construct ensures all flag values are valid and HTTP client/request built.
+func (i *imageSBOMCommand) construct(cobraCmd *cobra.Command) error {
+	var err error
+
+	// Validate the image reference.
+	if err := imageUtils.IsValidImageString(i.image); err != nil {
+		return common.ErrInvalidCommandOption.CausedBy(errors.Wrap(err, "image"))
+	}
+
+	// Create the HTTP client.
+	i.client, err = i.env.HTTPClient(flags.Timeout(cobraCmd))
```

Comment:
Is there a reason you are not using gRPC with typed request objects here? It seems like the gRPC handler was not implemented on the backend, but I didn't find an explanation in https://github.com/stackrox/stackrox/pull/13345

Reply:
The explanation and associated research is in the design document (I'll send you the link on slack), refer to appendix `Research: gRPC Gateway Returning Raw Response Bodies`.  

The reasoning is the same/similar to the other non-grpc endpoints, TL;DR; we need the raw bytes in the response - this allows support for the UI's native browser download functionality and makes API requests function conventionally with the appropriate content type.